File: 3.5/2.0/LinFu.DesignByContract2.Examples/SampleLibrary/IContractForBankAccount.vb
Imports LinFu.DesignByContract2.Injectors
Imports LibraryInterfaces
Imports VBContractAssertions

<ContractFor(GetType(IBankAccount)), _
ShouldAlwaysHaveAnOwner(), _
BalanceIsAlwaysNonNegative()> _
Public Interface IContractForBankAccount
    Property Balance() As Integer

    <EnsureBalanceReflectsDepositAmount()> _
    Sub Deposit(<NonNegative()> ByVal amount As Integer)

    <EnsureBalanceReflectsWithdrawalAmount(), MinorsCannotWithdrawAnyMoney()> _
    Sub Withdraw(<NonNegative()> ByVal amount As Integer)

End Interface

File: 3.5/2.0/LinFu.DesignByContract2.Examples/CSSampleLibrary/IContractForBankAccount.cs
using System;
using System.Collections.Generic;
using System.Text;
using CSContractAssertions;
using LibraryInterfaces;
using LinFu.DesignByContract2.Injectors;

namespace CSSampleLibrary
{
    [ContractFor(typeof(IBankAccount))]
    [type: ShouldAlwaysHaveAnOwner]
    [type: BalanceIsAlwaysNonNegative]
    public interface IContractForBankAccount
    {
        int Balance { [return: NonNegative] get; }        
        [EnsureBalanceReflectsDepositAmount] void Deposit([NonNegative] int amount);
        
        [EnsureBalanceReflectsWithdrawalAmount, 
        MinorsCannotWithdrawAnyMoney] void Withdraw([NonNegative] int amount);
    }
}
